max7219/max7219.go

```diff
@@ -0,0 +1,458 @@
+// Copyright 2024 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+//
+// The max7219 package provides a simple interface for displaying data on
+// numeric 7-segment displays, or on matrix displays. It simplifies writes
+// and provides useful features like scrolling characters on either type
+// of display unit.
+package max7219
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"periph.io/x/conn/v3/physic"
+	"periph.io/x/conn/v3/spi"
+)
+
+// DecodeMode is the mode for handling data. Refer to the datasheet for
+// more information.
+type DecodeMode byte
+
+const (
+	_REGISTER_NOOP         byte = 0x0
+	_REGISTER_DECODE_MODE  byte = 0x9
+	_REGISTER_INTENSITY    byte = 0xa
+	_REGISTER_SCAN_LIMIT   byte = 0xb
+	_REGISTER_SHUTDOWN     byte = 0xc
+	_REGISTER_DISPLAY_TEST byte = 0xf
+
+	// Value to write to a Code B Font decoded register to blank out the
+	// digit.
+	ClearDigit byte = 0x0f
+	// Value to write for a minus sign symbol
+	MinusSign byte = 0x0a
+	// To turn the decimal point on for a digit display, OR the value of the
+	// digit with DecimalPoint
+	DecimalPoint byte = 0x80
+
+	// DecodeB is used for numeric segment displays. E.G. given a binary 0,
+	// it would turn on the appropriate segments to display the character 0.
+	DecodeB DecodeMode = 0xff
+	// DecodeNone is RAW mode, or not decoded. For each byte, bits that are
+	// one are turned on in the matrix, and bits that are 0 turn off the
+	// led at that row/column.
+	DecodeNone DecodeMode = 0
+)
+
+// Type for a Maxim MAX7219/MAX7221 device.
+type Dev struct {
+	conn spi.Conn
+	mu   sync.Mutex
```

Comment:
I don't believe this device should be thread safe, I'd personally skip on the locking and would expect the caller to ensure locking. That's what most devices do.